lib/Mojo/WebSocketProxy/Dispatcher.pm

```diff
@@ -0,0 +1,220 @@
+
+    return;
+}
+
+sub on_message {
+    my ($c, $args) = @_;
+
+    my $config = Mojo::WebSocketProxy::Dispatcher::Config->new->{config};
+
+    my $result;
+    my $req_storage = {};
+    $req_storage->{args} = $args;
+    timeout 15 => sub {
+        $result = Mojo::WebSocketProxy::Dispatcher::Parser::parse_req($c, $req_storage);
+        if (!$result
+            && (my $action = $c->dispatch($args)))
+        {
+            %$req_storage = (%$req_storage, %$action);
+            $req_storage->{method} = $req_storage->{name};
+            $result = $c->before_forward($req_storage);
+
+            # Don't forward call to RPC if any before_forward hook returns response
+            # Or if there is instead_of_forward action
+            unless ($result) {
+                $result =
+                      $req_storage->{instead_of_forward}
+                    ? $req_storage->{instead_of_forward}->($c, $req_storage)
+                    : $c->forward($req_storage);
+            }
+
+            $c->after_forward($result, $req_storage);
+        } elsif (!$result) {
+            $c->app->log->debug("unrecognised request: " . $c->dumper($args));
+            $result = $c->wsp_error('error', 'UnrecognisedRequest', 'Unrecognised request.');
+        }
+    };
+    if ($@) {
+        $c->app->log->info("$$ timeout for " . JSON::to_json($args));
+    }
+
+    $c->send_api_response($req_storage, $result) if $result;
+
+    $c->_run_hooks($config->{after_dispatch} || []);
+
+    return;
+}
+
+sub before_forward {
+    my ($c, $req_storage) = @_;
+
+    my $config = Mojo::WebSocketProxy::Dispatcher::Config->new->{config};
+
+    # Should first call global hooks
+    my $before_forward_hooks = [
+        ref($config->{before_forward}) eq 'ARRAY'      ? @{$config->{before_forward}}             : $config->{before_forward},
+        ref($req_storage->{before_forward}) eq 'ARRAY' ? @{delete $req_storage->{before_forward}} : delete $req_storage->{before_forward},
+    ];
+
+    return $c->_run_hooks($before_forward_hooks, $req_storage);
+}
+
+sub after_forward {
+    my ($c, $result, $req_storage) = @_;
+
+    my $config = Mojo::WebSocketProxy::Dispatcher::Config->new->{config};
+    return $c->_run_hooks($config->{after_forward} || [], $result, $req_storage);
+}
+
+sub _run_hooks {
+    my @hook_params = @_;
+    my $c           = shift @hook_params;
+    my $hooks       = shift @hook_params;
+
+    my $i = 0;
+    my $result;
+    while (!$result && $i < @$hooks) {
+        my $hook = $hooks->[$i++];
+        next unless $hook;
+        $result = $hook->($c, @hook_params);
+    }
+
+    return $result;
+}
+
+sub dispatch {
+    my ($c, $args) = @_;
+
+    my $log = $c->app->log;
+    $log->debug("websocket got json " . $c->dumper($args));
+
+    my ($action) =
+        sort { $a->{order} <=> $b->{order} }
+        grep { defined }
+        map  { Mojo::WebSocketProxy::Dispatcher::Config->new->{actions}->{$_} } keys %$args;
```

Comment:
> why do that at runtime?

We should do that at runtime anyway, regardless of using array or hash. Because we should do that for every message. In the normal way sort won't be run, because only one action will be find for the message.

>  Why do not have actions in the config as array, and always pick the first?

 Because getting value by hash key is faster than grep whole array every time, I think.